# SETUP

In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
from copy import deepcopy
os.chdir(r'C:\Users\hitan\OneDrive\Desktop\MiniProjects\ML_NMIMS Codes')

In [5]:
summary_mean = pd.read_csv(r'CPCB_Issues\AirPy_v2\new_data\summary\summary_mean_AQI.csv')
summary_count = pd.read_csv(r'CPCB_Issues\AirPy_v2\new_data\summary\summary_count_AQI.csv')
sites_in_consideration = pd.read_csv(r'CPCB_Issues\AirPy_v2\final_site_list_new.csv')
summary_mean = deepcopy(summary_mean[summary_mean['site_id'].isin(sites_in_consideration['site_id'])].reset_index(drop=True))
summary_mean['pollutant'] = pd.merge(summary_mean, sites_in_consideration.groupby('site_id').agg({'pollutant':','.join}).reset_index(), on='site_id', how='inner')['pollutant']  
site_master = pd.read_csv(r'CPCB_Issues\AirPy_v2\sites_master.csv')
NO2_sites_mean = summary_mean[summary_mean['site_id'].isin(sites_in_consideration[sites_in_consideration['pollutant'] == 'NO2']['site_id'])]
PM10_sites_mean = summary_mean[summary_mean['site_id'].isin(sites_in_consideration[sites_in_consideration['pollutant'] == 'PM10']['site_id'])]
PM25_sites_mean = summary_mean[summary_mean['site_id'].isin(sites_in_consideration[sites_in_consideration['pollutant'] == 'PM25']['site_id'])]
Ozone_sites_mean = summary_mean[summary_mean['site_id'].isin(sites_in_consideration[sites_in_consideration['pollutant'] == 'Ozone']['site_id'])]
NO2_sites_count = summary_count[summary_count['site_id'].isin(sites_in_consideration[sites_in_consideration['pollutant'] == 'NO2']['site_id'])]
PM10_sites_count = summary_count[summary_count['site_id'].isin(sites_in_consideration[sites_in_consideration['pollutant'] == 'PM10']['site_id'])]
PM25_sites_count = summary_count[summary_count['site_id'].isin(sites_in_consideration[sites_in_consideration['pollutant'] == 'PM25']['site_id'])]
Ozone_sites_count = summary_count[summary_count['site_id'].isin(sites_in_consideration[sites_in_consideration['pollutant'] == 'Ozone']['site_id'])]

In [3]:
summary_mean_map_plot = pd.read_csv(r'CPCB_Issues\AirPy_v2\new_data\summary\summary_mean_map_plot.csv')

In [4]:
pols = ['PM25', 'PM10', 'NO2', 'NO', 'Ozone']

In [5]:
cols = summary_mean.columns.tolist()

#
Number of common sites 

In [9]:
len(summary_mean[summary_mean['pollutant'] == 'PM25,PM10,Ozone,NO2']['site_id'].unique())

144

#
We find that removing consecutively similar observations and outliers changes annual mean pollutant concentrations by only <5% but correcting for the yet unaddressed issue of unit inconsistency increases annual mean NO2 concentrations by 50-60%.

In [15]:
f1 = lambda x: x
f2 = lambda x: x + '_clean'

desc = deepcopy(summary_mean[[f(x) for x in pols for f in (f1,f2)]])
desc['site_id'] = summary_mean['site_id']
desc['year'] = summary_mean['year']
desc['pollutant'] = summary_mean['pollutant']
# desc = desc.describe().iloc[[1,2,3,-1],:]

for pol in pols:
    desc[pol + '_%change'] = (desc[pol + '_clean'] - desc[pol])/ desc[pol] * 100

# desc.iloc[desc[list(filter(lambda x: '%change' in x, desc.columns))].idxmax(axis=0)]

for pol in pols:
    print(pol)
    temp = desc[desc['pollutant'].str.contains(pol)]
    print(temp[pol + '_%change'].describe())
    # print(temp.loc[temp[pol + '_%change'].abs().idxmax()])
    print()

    


PM25
count    937.000000
mean      -1.226777
std        2.284149
min      -21.435097
25%       -1.727999
50%       -1.083614
75%       -0.580843
max       19.771061
Name: PM25_%change, dtype: float64

PM10
count    875.000000
mean      -1.360990
std        1.886255
min      -10.921732
25%       -1.933420
50%       -1.352093
75%       -0.871151
max       17.084470
Name: PM10_%change, dtype: float64

NO2
count    893.000000
mean       0.601576
std       12.566937
min      -73.731877
25%       -2.670542
50%       -1.518561
75%        0.193013
max      136.128016
Name: NO2_%change, dtype: float64

NO
count    891.000000
mean      -2.702990
std       18.726934
min      -61.792989
25%       -8.521560
50%       -4.322464
75%       -0.252558
max      381.430933
Name: NO_%change, dtype: float64

Ozone
count    907.000000
mean       0.604694
std        8.363500
min      -54.491140
25%       -0.928400
50%        0.114024
75%        0.912781
max      141.979801
Name: Ozone_%change, dtype: float64


In [11]:
pd.read_csv(r'CPCB_Issues\AirPy_v2\new_data\After_Cleaning_New\site_5267_Ibrahimpur_Vijayapura_KSPCB_2022.csv').describe().T['mean']

C:\Users\hitan\AppData\Local\Temp\ipykernel_15084\2142966233.py:1: DtypeWarning: Columns (32) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv(r'CPCB_Issues\AirPy_v2\new_data\After_Cleaning_New\site_5267_Ibrahimpur_Vijayapura_KSPCB_2022.csv').describe().T['mean']


state                         NaN
PM25                    16.858227
PM10                    49.024093
NO                       4.605940
NO2                      9.276348
NOx                     13.163947
NH3                      7.657276
SO2                      5.196426
CO                       0.600953
Ozone                   12.457677
PM25_consecutives       23.777744
PM25_outliers           23.976407
PM25_clean              23.976407
PM10_consecutives       53.529137
PM10_outliers           53.943958
PM10_clean              53.943958
NOx_consecutives        17.537946
NOx_outliers            17.486272
NOx_clean               17.486272
NO2_consecutives        18.029747
NO2_outliers            18.111600
NO2_clean               18.111600
NO_consecutives          4.052101
NO_outliers              4.039379
NO_clean                 4.039379
Ozone_consecutives      22.949473
Ozone_outliers          23.162787
Ozone_clean             23.162787
NO_CPCB                  4.620123
NO2_CPCB      

In [12]:
from copy import deepcopy
suffix = lambda x: '_CPCB' if x == 'NO2' else '_clean'
f1 = lambda x: x
f2 = lambda x: x + suffix(x)

desc = deepcopy(summary_mean[[f(x) for x in pols for f in (f1,f2)]])
desc['site_id'] = summary_mean['site_id']
desc['year'] = summary_mean['year']
desc['pollutant'] = summary_mean['pollutant']
# desc = desc.describe().iloc[[1,2,3,-1],:]

for pol in pols:
    desc[pol + '_%change'] = (desc[pol + suffix(pol)] - desc[pol])/ desc[pol] * 100

# desc.iloc[desc[list(filter(lambda x: 'change' in x, desc.columns))].idxmax(axis=0)]

for pol in pols:
    print(pol)
    temp = desc[desc['pollutant'].str.contains(pol)]
    print(temp[pol + '_%change'].describe()[['max']])
    print(temp.loc[temp[pol + '_%change'].abs().idxmax()])
    print()

    


PM25
max    3.805875
Name: PM25_%change, dtype: float64
PM25             60.989142
PM25_clean       55.801593
PM10                   NaN
PM10_clean             NaN
NO2              22.311077
NO2_CPCB          35.75897
NO               13.035476
NO_clean         12.373498
Ozone                  NaN
Ozone_clean            NaN
site_id           site_306
year                  2019
pollutant         PM25,NO2
PM25_%change     -8.505694
PM10_%change           NaN
NO2_%change      60.274512
NO_%change       -5.078287
Ozone_%change          NaN
Name: 425, dtype: object

PM10
max    13.462661
Name: PM10_%change, dtype: float64
PM25                       30.594982
PM25_clean                 30.374759
PM10                       79.125834
PM10_clean                 89.778278
NO2                        15.196031
NO2_CPCB                    5.627597
NO                          9.545288
NO_clean                   10.971903
Ozone                      29.743233
Ozone_clean                34.278031
site_

#
15 out of 40 sites became non-compliant to national ambient air quality standards, the Air Quality Index increased significantly by 11.4%, NOx-to-NO2 declined by 20% on average across all site

In [18]:
print('Original Number of sites: ', summary_mean_map_plot['site_id'].nunique())
print('Number of sites with compliance change: ', summary_mean_map_plot[summary_mean_map_plot['compliance_change_compl_to_non']]['site_id'].nunique())

Original Number of sites:  216
Number of sites with compliance change:  90


#
The number of Continuous Ambient Air Quality Monitoring Sites (CAAQMS) maintained by the Central Pollution Control Board (CPCB) and other non-CPCB organizations such as State Pollution Control Boards (SPCBs), Indian Institute of Tropical Meteorology (IITM) and India Meteorological Department (IMD) has increased from 30 in 2013 to more than 500 in 2023

In [35]:
print('Number of sites: ', len(site_master))

Number of sites:  539


#
were set up in 2019, and this, in addition to our site selection criterion, reduces bring the number of sites to 840 or more for each of the pollutants

In [6]:
sites_in_consideration
print(sites_in_consideration.groupby(['pollutant']).agg({'site_id':'count'}))

           site_id
pollutant         
NO2            184
Ozone          188
PM10           182
PM25           193


#
Only 11 out of 40 sites consistently adhered to the protocol,

In [7]:
NO2_sites = summary_mean[summary_mean['site_id'].isin(sites_in_consideration[sites_in_consideration['pollutant'] == 'NO2']['site_id'])]

In [21]:
NO2_sites['site_id'].nunique()

184

In [20]:
NO2_sites[NO2_sites['prevalent_error'] == 'C1'].groupby('site_id').agg({'site_id':'count'})['site_id'].value_counts()

5    46
3    20
2    17
4    12
1    10
Name: site_id, dtype: int64

 # 
 and 12 sites continuously misreported both NO and NO2 levels in ppb during 2019 and 2023

In [25]:
NO2_sites[NO2_sites['prevalent_error'] == 'C2'].groupby('site_id').agg({'site_id':'count'})['site_id'].value_counts()


5    46
3    13
4    10
1     9
2     8
Name: site_id, dtype: int64

We observed 9.2% of trace gases and 2.1% of PM datasets as consecutive similar (Medians)

In [26]:
medians = summary_mean[summary_mean['site_id'].isin(sites_in_consideration[sites_in_consideration['pollutant'] == 'NO2']['site_id'])]

timestamp  state        PM25        PM10         NO        NO2  \
0       17519.5    NaN  105.092793  212.100545  64.806904  51.883613   
1       17567.5    NaN   93.382654  196.633830  60.781706  37.409129   
2       17519.5    NaN  106.071454  235.346037  79.917620  50.195624   
3       17519.5    NaN   87.775139  204.064709  45.562646  13.807312   
4       17519.5    NaN   84.599676  179.539458  65.525145  21.553671   
...         ...    ...         ...         ...        ...        ...   
1068    17519.5    NaN         NaN         NaN        NaN        NaN   
1069    17567.5    NaN         NaN         NaN        NaN        NaN   
1070    17519.5    NaN   20.904096   43.553922   2.232706  11.486363   
1071    17519.5    NaN   25.434799   52.056467   2.321842  11.002988   
1072    17519.5    NaN   25.002744   52.650156   2.610839  10.750275   

            NOx       NH3        SO2        CO  ...  year    site_id  \
0     94.403413       NaN        NaN  1.565263  ...  2019   site_103   
1     92.122695       NaN        NaN  1.367190  ...  2020   site_103   
2     81.560752       NaN        NaN  2.312735  ...  2021   site_103   
3     56.262578       NaN        NaN  1.900407  ...  2022   site_103   
4     78.309509       NaN        NaN  2.016638  ...  2023   site_103   
...         ...       ...        ...       ...  ...   ...        ...   
1068        NaN       NaN        NaN       NaN  ...  2019  site_5423   
1069        NaN       NaN        NaN       NaN  ...  2020  site_5423   
1070   7.827759  8.800420   8.799872  0.530921  ...  2021  site_5423   
1071   7.690184  9.086578  12.814797  0.559393  ...  2022  site_5423   
1072   7.824315  9.233723   8.201797  0.703191  ...  2023  site_5423   

                          site_name  prevalent_error  \
0       CRRI_Mathura_Road_Delhi_IMD               C2   
1       CRRI_Mathura_Road_Delhi_IMD               C2   
2       CRRI_Mathura_Road_Delhi_IMD               C2   
3       CRRI_Mathura_Road_Delhi_IMD               C2   
4       CRRI_Mathura_Road_Delhi_IMD               C2   
...                             ...              ...   
1068  Jawahar_Nagar_Puducherry_PPCC              NaN   
1069  Jawahar_Nagar_Puducherry_PPCC              NaN   
1070  Jawahar_Nagar_Puducherry_PPCC               C1   
1071  Jawahar_Nagar_Puducherry_PPCC               C1   
1072  Jawahar_Nagar_Puducherry_PPCC               C1   

                          errors  C3  C5  C7  C8            pollutant  
0     C1,C2,C4,C6,UNIDENTIFIABLE NaN NaN NaN NaN  PM25,PM10,Ozone,NO2  
1     C1,C2,C4,C6,UNIDENTIFIABLE NaN NaN NaN NaN  PM25,PM10,Ozone,NO2  
2     C1,C2,C4,C6,UNIDENTIFIABLE NaN NaN NaN NaN  PM25,PM10,Ozone,NO2  
3     C1,C2,C4,C6,UNIDENTIFIABLE NaN NaN NaN NaN  PM25,PM10,Ozone,NO2  
4     C1,C2,C4,C6,UNIDENTIFIABLE NaN NaN NaN NaN  PM25,PM10,Ozone,NO2  
...                          ...  ..  ..  ..  ..                  ...  
1068                         NaN NaN NaN NaN NaN  PM25,PM10,Ozone,NO2  
1069                         NaN NaN NaN NaN NaN  PM25,PM10,Ozone,NO2  
1070                       C1,C4 NaN NaN NaN NaN  PM25,PM10,Ozone,NO2  
1071                    C1,C4,C6 NaN NaN NaN NaN  PM25,PM10,Ozone,NO2  
1072     C1,C4,C6,UNIDENTIFIABLE NaN NaN NaN NaN  PM25,PM10,Ozone,NO2  

[1073 rows x 48 columns]

#
 We observed 9.2% of trace gases and 2.1% of PM datasets as consecutive similar

In [32]:
NO2_sites.columns

Index(['timestamp', 'state', 'PM25', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'SO2',
       'CO', 'Ozone', 'PM25_consecutives', 'PM25_outliers', 'PM25_clean',
       'PM10_consecutives', 'PM10_outliers', 'PM10_clean', 'NOx_consecutives',
       'NOx_outliers', 'NOx_clean', 'NO2_consecutives', 'NO2_outliers',
       'NO2_clean', 'NO_consecutives', 'NO_outliers', 'NO_clean',
       'Ozone_consecutives', 'Ozone_outliers', 'Ozone_clean', 'NO_CPCB',
       'NO2_CPCB', 'NOx_CPCB', 'Threshold', 'C1', 'C2', 'C4', 'C6', 'mismatch',
       'year', 'site_id', 'site_name', 'prevalent_error', 'errors', 'C3', 'C5',
       'C7', 'C8', 'pollutant'],
      dtype='object')

In [56]:
print('NO2')
NO2_sites_count_groups = NO2_sites_count.groupby('year')
for year in NO2_sites_count_groups.groups.keys():
    NO2 = NO2_sites_count_groups.get_group(year)
    # print(NO2)
    print(year, ((NO2['NO2'] - NO2['NO2_consecutives'])/NO2['NO2'] * 100).mean(),'%')
print()

print('PM10')
PM10_sites_count_groups = PM10_sites_count.groupby('year')
for year in PM10_sites_count_groups.groups.keys():
    PM10 = PM10_sites_count_groups.get_group(year)
    print(year, ((PM10['PM10'] - PM10['PM10_consecutives'])/PM10['PM10'] * 100).mean(),'%')
print()

print('PM25')
PM25_sites_count_groups = PM25_sites_count.groupby('year')
for year in PM25_sites_count_groups.groups.keys():
    PM25 = PM25_sites_count_groups.get_group(year)
    print(year, ((PM25['PM25'] - PM25['PM25_consecutives'])/PM25['PM25'] * 100).median(),'%')
print()

print('Ozone')
Ozone_sites_count_groups = Ozone_sites_count.groupby('year')
for year in Ozone_sites_count_groups.groups.keys():
    Ozone = Ozone_sites_count_groups.get_group(year)
    print(year, ((Ozone['Ozone'] - Ozone['Ozone_consecutives'])/Ozone['Ozone'] * 100).median(),'%')
print()





NO2
2019 0.24874669932264362 %
2020 0.4364544460026524 %
2021 2.2422665678203653 %
2022 4.964300111896601 %
2023 6.158796937819098 %

PM10
2019 0.1283190719353656 %
2020 0.2026178222636463 %
2021 0.17998796247299456 %
2022 0.17854653243809684 %
2023 0.30105524518932336 %

PM25
2019 0.09359498330889464 %
2020 0.12252050658464593 %
2021 0.13689907672715695 %
2022 0.15661466650288663 %
2023 0.172536548757961 %

Ozone
2019 0.14612761811982464 %
2020 0.1970018776741466 %
2021 0.2501453101632489 %
2022 0.4709504140572166 %
2023 1.118734161567464 %



In [54]:
print('NO2')
NO2_sites_count_groups = NO2_sites_count.groupby('year')
for year in NO2_sites_count_groups.groups.keys():
    NO2 = NO2_sites_count_groups.get_group(year)
    print(year, ((NO2['NO2_consecutives'] - NO2['NO2_outliers'])/NO2['NO2'] * 100).median(),'%')
print()

print('PM10')
PM10_sites_count_groups = PM10_sites_count.groupby('year')
for year in PM10_sites_count_groups.groups.keys():
    PM10 = PM10_sites_count_groups.get_group(year)
    print(year, ((PM10['PM10_consecutives'] - PM10['PM10_outliers'])/PM10['PM10'] * 100).median(),'%')
print()

print('PM25')
PM25_sites_count_groups = PM25_sites_count.groupby('year')
for year in PM25_sites_count_groups.groups.keys():
    PM25 = PM25_sites_count_groups.get_group(year)
    print(year, ((PM25['PM25_consecutives'] - PM25['PM25_outliers'])/PM25['PM25'] * 100).median(),'%')
print()

print('Ozone')
Ozone_sites_count_groups = Ozone_sites_count.groupby('year')
for year in Ozone_sites_count_groups.groups.keys():
    Ozone = Ozone_sites_count_groups.get_group(year)
    print(year, ((Ozone['Ozone_consecutives'] - Ozone['Ozone_outliers'])/Ozone['Ozone'] * 100).median(),'%')
print()

NO2
2019 8.125067618738505 %
2020 7.906860991591926 %
2021 7.54509058314558 %
2022 7.366651568576101 %
2023 7.321254075646712 %

PM10
2019 10.342899596956844 %
2020 9.845266990291263 %
2021 9.86304639440101 %
2022 9.832814682326404 %
2023 10.202381365640296 %

PM25
2019 10.149647887323944 %
2020 9.693230278994903 %
2021 9.662294874231767 %
2022 9.683338078516972 %
2023 9.948773717005025 %

Ozone
2019 8.938153064106185 %
2020 8.720793013924947 %
2021 8.626235735438527 %
2022 8.873596432627256 %
2023 8.384240814519698 %



#
In addition to inconsistency in units, consecutive similar observations, and outliers, air quality data is also affected by minor issues such as PM2.5 exceeding PM10 and negative observations. Overall, 28 out of 40 sites reported PM2.5 concentrations exceeding PM10.

In [104]:
import tqdm
import glob
files = glob.glob('CPCB_Issues/AirPy_v2/new_data/After_Cleaning_New/*')
def summarize(files: list):
    summaries = []
    l = []
    usecols = ['dates', 'site_id', 'PM25', 'PM10', 'year']
    cols = pd.read_csv(files[0], usecols=usecols).columns
    for file in tqdm.tqdm(files):
        try:
            df = pd.read_csv(file, usecols=usecols)
            PM_exceeding = df[df['PM25'] > df['PM10']]
            if len(PM_exceeding) > 0:
               l.extend(PM_exceeding.values.tolist())
        except:
            print('ERROR: ', file)
            continue
    df = pd.DataFrame(l, columns=usecols).to_csv('CPCB_Issues/AirPy_v2/new_data/summary/PM25_gt_PM10.csv', index=False)
    return df


if __name__ == '__main__':
    print(summarize(files))

 27%|██▋       | 288/1080 [00:21<00:58, 13.65it/s]

ERROR:  CPCB_Issues/AirPy_v2/new_data/After_Cleaning_New\site_147_MD_University_Rohtak_HSPCB_2019.csv


 35%|███▌      | 381/1080 [00:28<00:43, 16.23it/s]

ERROR:  CPCB_Issues/AirPy_v2/new_data/After_Cleaning_New\site_158_Muzaffarpur_Collectorate_Muzaffarpur_BSPCB_2022.csv


100%|██████████| 1080/1080 [01:15<00:00, 14.32it/s]


None


In [101]:
exceeding_sites = pd.read_csv(r'CPCB_Issues\AirPy_v2\new_data\summary\PM25_gt_PM10.csv')

In [103]:
exceeding_sites['site_id']

0        247.76
1        160.67
2        168.87
3        162.99
4        154.69
          ...  
33734     19.99
33735     26.00
33736     15.00
33737      8.00
33738      8.00
Name: site_id, Length: 33739, dtype: float64

#
Annual mean pollutant concentrations across all sites increased by 2-4% on median for trace gases and by 0.01-0.04% for PM after removing consecutive similar observations 

In [18]:
for pol in ['PM25', 'PM10', 'NO2', 'Ozone']:
    s_mean = summary_mean[summary_mean['site_id'].isin(sites_in_consideration[sites_in_consideration['pollutant'] == pol]['site_id'])]
    increase = ((s_mean[pol + '_consecutives'] - s_mean[pol])/s_mean[pol]) * 100
    print(pol)
    print(increase.min(),increase.max(),increase.median())
print()

PM25
-21.19222137355666 20.904505088432945 -0.04628116141306543
PM10
-8.585993292029254 19.00332544046779 -0.01748523636173128
NO2
-73.88304089575692 136.3438104781974 0.022633280425061627
Ozone
-73.58642850878572 76.45890445493654 0.006225013191343985



In [25]:
print('NO2')
print(NO2_sites_mean['NO2'].min(), NO2_sites_mean['NO2'].max(), NO2_sites_mean['NO2'].mean())
print(NO2_sites_mean['NO2_consecutives'].min(), NO2_sites_mean['NO2_consecutives'].max(), NO2_sites_mean['NO2_consecutives'].median())
print((NO2_sites_mean['NO2_consecutives'].median() - NO2_sites_mean['NO2'].median())/NO2_sites_mean['NO2'].median() * 100)
print()
print('PM10')
print(PM10_sites_mean['PM10'].min(), PM10_sites_mean['PM10'].max(), PM10_sites_mean['PM10'].median())
print(PM10_sites_mean['PM10_consecutives'].min(), PM10_sites_mean['PM10_consecutives'].max(), PM10_sites_mean['PM10_consecutives'].median())
print((PM10_sites_mean['PM10_consecutives'].median() - PM10_sites_mean['PM10'].median())/PM10_sites_mean['PM10'].median() * 100)
print()
print('PM25')
print(PM25_sites_mean['PM25'].min(), PM25_sites_mean['PM25'].max(), PM25_sites_mean['PM25'].median())
print(PM25_sites_mean['PM25_consecutives'].min(), PM25_sites_mean['PM25_consecutives'].max(), PM25_sites_mean['PM25_consecutives'].median())
print((PM25_sites_mean['PM25_consecutives'].median() - PM25_sites_mean['PM25'].median())/PM25_sites_mean['PM25'].median() * 100)
print()
print('Ozone')
print(Ozone_sites_mean['Ozone'].min(), Ozone_sites_mean['Ozone'].max(), Ozone_sites_mean['Ozone'].median())
print(Ozone_sites_mean['Ozone_consecutives'].min(), Ozone_sites_mean['Ozone_consecutives'].max(), Ozone_sites_mean['Ozone_consecutives'].median())
print((Ozone_sites_mean['Ozone_consecutives'].median() - Ozone_sites_mean['Ozone'].median())/Ozone_sites_mean['Ozone'].median() * 100)
print()


NO2
2.5545057747552327 107.81492512199227 28.90754272555247
1.635593705293276 108.8828167438888 25.73803146573435
2.0436626947746035

PM10
29.02245043994488 316.7321611061016 124.51104694312835
29.02149663112101 316.49624127976904 125.30989303146744
0.6415865161779346

PM25
10.2881738680168 160.61694444444444 60.88691868017667
10.461499645054426 161.26559646539027 60.77421139134592
-0.18510920124365385

Ozone
0.4011911546791291 93.32206084147094 30.840763459158417
0.409699814217268 91.4798455311973 30.813491960091884
-0.08842679625181042



In [26]:
print('NO2')
print(NO2_sites_mean['NO2'].min(), NO2_sites_mean['NO2'].max(), NO2_sites_mean['NO2'].mean())
print(NO2_sites_mean['NO2_consecutives'].min(), NO2_sites_mean['NO2_consecutives'].max(), NO2_sites_mean['NO2_consecutives'].mean())
print((NO2_sites_mean['NO2_consecutives'].mean() - NO2_sites_mean['NO2'].mean())/NO2_sites_mean['NO2'].mean() * 100)
print()
print('PM10')
print(PM10_sites_mean['PM10'].min(), PM10_sites_mean['PM10'].max(), PM10_sites_mean['PM10'].mean())
print(PM10_sites_mean['PM10_consecutives'].min(), PM10_sites_mean['PM10_consecutives'].max(), PM10_sites_mean['PM10_consecutives'].mean())
print((PM10_sites_mean['PM10_consecutives'].mean() - PM10_sites_mean['PM10'].mean())/PM10_sites_mean['PM10'].mean() * 100)
print()
print('PM25')
print(PM25_sites_mean['PM25'].min(), PM25_sites_mean['PM25'].max(), PM25_sites_mean['PM25'].mean())
print(PM25_sites_mean['PM25_consecutives'].min(), PM25_sites_mean['PM25_consecutives'].max(), PM25_sites_mean['PM25_consecutives'].mean())
print((PM25_sites_mean['PM25_consecutives'].mean() - PM25_sites_mean['PM25'].mean())/PM25_sites_mean['PM25'].mean() * 100)
print()
print('Ozone')
print(Ozone_sites_mean['Ozone'].min(), Ozone_sites_mean['Ozone'].max(), Ozone_sites_mean['Ozone'].mean())
print(Ozone_sites_mean['Ozone_consecutives'].min(), Ozone_sites_mean['Ozone_consecutives'].max(), Ozone_sites_mean['Ozone_consecutives'].mean())
print((Ozone_sites_mean['Ozone_consecutives'].mean() - Ozone_sites_mean['Ozone'].mean())/Ozone_sites_mean['Ozone'].mean() * 100)
print()

NO2
2.5545057747552327 107.81492512199227 28.90754272555247
1.635593705293276 108.8828167438888 29.16851360700598
0.9027778110756803

PM10
29.02245043994488 316.7321611061016 136.3564158305143
29.02149663112101 316.49624127976904 136.5194475364827
0.1195629153020825

PM25
10.2881738680168 160.61694444444444 66.27826812951069
10.461499645054426 161.26559646539027 66.22235866735511
-0.0843556473840465

Ozone
0.4011911546791291 93.32206084147094 31.542287096812046
0.409699814217268 91.4798455311973 31.580472361933627
0.12106054644794452



In [27]:
print('NO2')
print(NO2_sites_mean['NO2'].min(), NO2_sites_mean['NO2'].max(), NO2_sites_mean['NO2'].mean())
print(NO2_sites_mean['NO2_outliers'].min(), NO2_sites_mean['NO2_outliers'].max(), NO2_sites_mean['NO2_outliers'].mean())
print((NO2_sites_mean['NO2_outliers'].mean() - NO2_sites_mean['NO2'].mean())/NO2_sites_mean['NO2'].mean() * 100)
print()
print('PM10')
print(PM10_sites_mean['PM10'].min(), PM10_sites_mean['PM10'].max(), PM10_sites_mean['PM10'].mean())
print(PM10_sites_mean['PM10_outliers'].min(), PM10_sites_mean['PM10_outliers'].max(), PM10_sites_mean['PM10_outliers'].mean())
print((PM10_sites_mean['PM10_outliers'].mean() - PM10_sites_mean['PM10'].mean())/PM10_sites_mean['PM10'].mean() * 100)
print()
print('PM25')
print(PM25_sites_mean['PM25'].min(), PM25_sites_mean['PM25'].max(), PM25_sites_mean['PM25'].mean())
print(PM25_sites_mean['PM25_outliers'].min(), PM25_sites_mean['PM25_outliers'].max(), PM25_sites_mean['PM25_outliers'].mean())
print((PM25_sites_mean['PM25_outliers'].mean() - PM25_sites_mean['PM25'].mean())/PM25_sites_mean['PM25'].mean() * 100)
print()
print('Ozone')
print(Ozone_sites_mean['Ozone'].min(), Ozone_sites_mean['Ozone'].max(), Ozone_sites_mean['Ozone'].mean())
print(Ozone_sites_mean['Ozone_outliers'].min(), Ozone_sites_mean['Ozone_outliers'].max(), Ozone_sites_mean['Ozone_outliers'].mean())
print((Ozone_sites_mean['Ozone_outliers'].mean() - Ozone_sites_mean['Ozone'].mean())/Ozone_sites_mean['Ozone'].mean() * 100)
print()

NO2
2.5545057747552327 107.81492512199227 28.90754272555247
2.404149809979159 111.02245221527347 28.822930462457705
-0.2926996040378536

PM10
29.02245043994488 316.7321611061016 136.3564158305143
28.452250489236796 312.6496918258058 134.62577202861695
-1.2692059932467472

PM25
10.2881738680168 160.61694444444444 66.27826812951069
9.047212269622374 160.88782537067547 65.51275187439384
-1.1550034071816682

Ozone
0.4011911546791291 93.32206084147094 31.542287096812046
0.401903776513192 99.60549822923376 31.65776759599136
0.36611327144690253



#
Resolving unit inconsistency increased annual mean NO2 by 50-60% and NO by 10-20% at 12 sites that continuously misreported both species in ppb instead of µg m-³ over the three-year perio

In [30]:
print('NO2')
print(NO2_sites_mean['NO2'].min(), NO2_sites_mean['NO2'].max(), NO2_sites_mean['NO2'].mean())
print(NO2_sites_mean['NO2_CPCB'].min(), NO2_sites_mean['NO2_CPCB'].max(), NO2_sites_mean['NO2_CPCB'].mean())
inc = (NO2_sites_mean['NO2_CPCB'].mean() - NO2_sites_mean['NO2'].mean())/NO2_sites_mean['NO2'] * 100
print(inc.min(), inc.max(), inc.mean())

NO2
2.5545057747552327 107.81492512199227 28.90754272555247
0.0439509132420091 130.2572151145492 36.31574241891968
6.871219068217926 290.00520439524354 36.287798821382395


In [32]:
print('NO')
print(NO2_sites_mean['NO'].min(), NO2_sites_mean['NO'].max(), NO2_sites_mean['NO'].mean())
print(NO2_sites_mean['NO_CPCB'].min(), NO2_sites_mean['NO_CPCB'].max(), NO2_sites_mean['NO_CPCB'].mean())
inc = (NO2_sites_mean['NO_CPCB'].mean() - NO2_sites_mean['NO'].mean())/NO2_sites_mean['NO'] * 100
print(inc.min(), inc.max(), inc.mean())

NO
0.3385948780820051 101.25040008573266 17.477442019424043
0.0 146.46678888888889 18.227313376991336
0.7406107599894405 221.46565294046786 9.265827604635117


#
After removing spurious outliers, consecutive similar observations, negative observations and instances of PM2.5 exceeding PM10 and addressing the novel issue of unit inconsistency, mean NO2 across all sites for 2019-2021 increased dramatically by 89% from 30.9 µg m-3 to 58.4 µg m-3. O3 had a small increase of 2.3% from 46.8 µg m-3 to 47.9 µg m-3 and PM had a marginal decline of 0.1% from 57.88 µg m-3 to 57.82 µg m-3 for PM2.5 and 0.7% from 111.95 µg m-3 to 111.76 µg m-3 for PM10.

In [72]:
import pandas as pd

def generate_summary(dataframe, column_name, clean_column_name, year):
    dataframe = dataframe[dataframe['year'] == year]
    summary = {
        'Metric': ['Min','25%ile','Mean', 'Median', '75%ile', 'Max', '% diff',],
        column_name: [
            dataframe[column_name].min(),
            dataframe[column_name].quantile(0.25),
            dataframe[column_name].mean(),
            dataframe[column_name].median(),
            dataframe[column_name].quantile(0.75),
            dataframe[column_name].max(),
            None
        ],
        clean_column_name: [
            dataframe[clean_column_name].min(),
            dataframe[clean_column_name].quantile(0.25),
            dataframe[clean_column_name].mean(),
            dataframe[clean_column_name].median(),
            dataframe[clean_column_name].quantile(0.75),
            dataframe[clean_column_name].max(),
            None
        ]
    }
    percentage_diff = ((dataframe[clean_column_name] - dataframe[column_name]) / dataframe[column_name] * 100).mean()
    summary[clean_column_name][-1] = percentage_diff
    return pd.DataFrame(summary)

def print_summary_tables(year):
    merged_summary = pd.DataFrame()
    # print(NO2_sites_mean.groupby('year')['year'].groups.get(year))
    no2_summary = generate_summary(NO2_sites_mean, 'NO2', 'NO2_consecutives', year)
    pm10_summary = generate_summary(PM10_sites_mean, 'PM10', 'PM10_consecutives', year)
    pm25_summary = generate_summary(PM25_sites_mean, 'PM25', 'PM25_consecutives', year)
    ozone_summary = generate_summary(Ozone_sites_mean, 'Ozone', 'Ozone_consecutives', year)

    no2_summary1 = generate_summary(NO2_sites_mean, 'NO2_consecutives', 'NO2_outliers', year).drop('NO2_consecutives', axis=1)
    pm10_summary1 = generate_summary(PM10_sites_mean, 'PM10_consecutives', 'PM10_outliers', year).drop('PM10_consecutives', axis=1)
    pm25_summary1 = generate_summary(PM25_sites_mean, 'PM25_consecutives', 'PM25_outliers', year).drop('PM25_consecutives', axis=1)
    ozone_summary1 = generate_summary(Ozone_sites_mean, 'Ozone_consecutives', 'Ozone_outliers', year).drop('Ozone_consecutives', axis=1)


    merged_summary = pd.merge(no2_summary, no2_summary1, on='Metric').merge(pm10_summary, on='Metric').merge(pm10_summary1, on='Metric').merge(pm25_summary, on='Metric').merge(pm25_summary1, on='Metric').merge(ozone_summary, on='Metric').merge(ozone_summary1, on='Metric')
    merged_summary['year'] = year
    return merged_summary

pd.concat([print_summary_tables(year) for year in range(2019, 2023)])




Metric         NO2  NO2_consecutives  NO2_outliers        PM10  \
0     Min    6.899777          6.898164      6.372836   29.022450   
1  25%ile   20.438168         19.869189     19.689754   99.492924   
2    Mean   33.971319         34.178785     33.607124  147.410754   
3  Median   32.407971         32.419285     31.798909  135.371034   
4  75%ile   43.461436         43.838944     43.074484  203.440748   
5     Max   90.804365         90.797483     88.687064  290.983326   
6  % diff         NaN          0.551409     -1.697123         NaN   
0     Min    2.675933          2.701816      2.677115   35.649819   
1  25%ile   16.416617         16.648414     16.728333   82.499759   
2    Mean   27.667936         27.994724     27.587421  125.474169   
3  Median   24.864333         25.427126     24.568780  117.332179   
4  75%ile   37.090366         38.073716     37.614190  166.055501   
5     Max   69.355116         64.526424     64.075565  246.038874   
6  % diff         NaN          2.147123     -0.570997         NaN   
0     Min    2.660870          2.655636      2.404150   37.203454   
1  25%ile   18.220468         18.115758     17.463731   91.436250   
2    Mean   29.178323         29.286324     28.842231  137.862194   
3  Median   25.701437         26.119121     25.716366  127.391726   
4  75%ile   36.622812         36.893396     35.986038  189.827867   
5     Max  107.814925        108.882817    111.022452  277.655634   
6  % diff         NaN          0.735269     -1.099594         NaN   
0     Min    2.554506          1.635594      2.583542   34.403648   
1  25%ile   17.281632         17.021745     17.091304   95.401385   
2    Mean   27.701975         27.972869     27.689152  139.268665   
3  Median   23.997623         23.723976     23.388193  131.619180   
4  75%ile   35.463909         35.464280     34.558672  184.174419   
5     Max   95.718501         95.842173     93.052928  316.732161   
6  % diff         NaN          2.241015     -0.427614         NaN   

   PM10_consecutives  PM10_outliers        PM25  PM25_consecutives  \
0          29.021497      28.452250   10.288174          10.461500   
1          99.435916      97.438084   44.212437          43.954819   
2         147.369067     145.283935   74.233101          74.163426   
3         135.292971     132.933442   73.960254          73.809242   
4         203.263529     201.969235  104.473113         104.333173   
5         290.954171     285.019405  160.616944         161.265596   
6          -0.019593      -1.488831         NaN          -0.109344   
0          35.654254      34.538528   11.343268          11.283058   
1          82.578684      81.982545   39.922556          39.869407   
2         125.586697     123.881278   63.590159          63.567090   
3         116.920268     115.446992   57.171098          57.382749   
4         165.670269     164.316457   88.629194          88.632541   
5         248.188406     245.726335  149.382414         149.615692   
6           0.137821      -1.477713         NaN          -0.010157   
0          36.880351      36.389216   17.613499          17.596136   
1          91.001948      89.481764   43.133566          43.110390   
2         137.743969     135.688202   67.175478          67.139591   
3         127.847723     125.840577   60.678775          60.652143   
4         189.598367     186.448161   93.606063          93.440730   
5         277.417145     274.753763  140.757038         140.428266   
6          -0.079669      -1.526397         NaN           0.072195   
0          34.109888      33.615193   17.104526          17.084871   
1          95.404644      94.429544   42.472130          42.635474   
2         139.378871     137.432134   64.329358          64.338167   
3         132.367286     130.592143   60.000401          59.969844   
4         184.096577     181.810735   87.775139          87.718469   
5         316.496241     312.649692  128.066366         127.828417   
6           0.093222      -1.437183      

In [8]:
print('NO2')
print(NO2_sites_mean['NO2'].min(), NO2_sites_mean['NO2'].max(), NO2_sites_mean['NO2'].mean())
print(NO2_sites_mean['NO2_CPCB'].min(), NO2_sites_mean['NO2_CPCB'].max(), NO2_sites_mean['NO2_CPCB'].mean())
print((NO2_sites_mean['NO2_CPCB'].mean() - NO2_sites_mean['NO2'].mean())/NO2_sites_mean['NO2'].mean() * 100)
print()
print('PM10')
print(PM10_sites_mean['PM10'].min(), PM10_sites_mean['PM10'].max(), PM10_sites_mean['PM10'].mean())
print(PM10_sites_mean['PM10_clean'].min(), PM10_sites_mean['PM10_clean'].max(), PM10_sites_mean['PM10_clean'].mean())
print((PM10_sites_mean['PM10_clean'].mean() - PM10_sites_mean['PM10'].mean())/PM10_sites_mean['PM10'].mean() * 100)
print()
print('PM25')
print(PM25_sites_mean['PM25'].min(), PM25_sites_mean['PM25'].max(), PM25_sites_mean['PM25'].mean())
print(PM25_sites_mean['PM25_clean'].min(), PM25_sites_mean['PM25_clean'].max(), PM25_sites_mean['PM25_clean'].mean())
print((PM25_sites_mean['PM25_clean'].mean() - PM25_sites_mean['PM25'].mean())/PM25_sites_mean['PM25'].mean() * 100)
print()
print('Ozone')
print(Ozone_sites_mean['Ozone'].min(), Ozone_sites_mean['Ozone'].max(), Ozone_sites_mean['Ozone'].mean())
print(Ozone_sites_mean['Ozone_clean'].min(), Ozone_sites_mean['Ozone_clean'].max(), Ozone_sites_mean['Ozone_clean'].mean())
print((Ozone_sites_mean['Ozone_clean'].mean() - Ozone_sites_mean['Ozone'].mean())/Ozone_sites_mean['Ozone'].mean() * 100)
print()

NO2
2.5545057747552327 107.81492512199227 28.90754272555247
0.0439509132420091 130.2572151145492 36.31574241891968
25.627220423750597

PM10
29.02245043994488 316.7321611061016 136.3564158305143
28.452250489236796 312.6496918258058 134.62577202861695
-1.2692059932467472

PM25
10.2881738680168 160.61694444444444 66.27826812951069
9.047212269622374 160.88782537067547 65.51275187439384
-1.1550034071816682

Ozone
0.4011911546791291 93.32206084147094 31.542287096812046
0.401903776513192 99.60549822923376 31.65776759599136
0.36611327144690253



#
After data cleaning, 2020 annual mean NO2 concentrations double in more than half of the 40 sites we study. 2020 O3 concentrations double only in two sites, both in Mumbai. 

In [15]:
# Show year by year difference between cleaned and original data
for pol in ['PM25', 'PM10', 'NO2', 'Ozone']:
    s_mean = summary_mean[summary_mean['site_id'].isin(sites_in_consideration[sites_in_consideration['pollutant'] == pol]['site_id'])]
    for year in s_mean['year'].unique():
        increase = ((s_mean[s_mean['year'] == year][pol + '_clean'] - s_mean[s_mean['year'] == year][pol])/s_mean[s_mean['year'] == year][pol]) * 100
        print(pol, year)
        print(increase.min(),increase.max(),increase.median(),increase.mean())

PM25 2019
-15.125493944640734 4.037606339369802 -1.1721397881733182 -1.559584669802975
PM25 2020
-8.96491933092493 7.743113114372401 -1.1048720246669883 -1.2289763079769143
PM25 2021
-11.934912018097418 19.771060537050726 -1.112674515505111 -1.061419771799852
PM25 2022
-19.642637456887986 11.331092157187157 -1.0250470612486853 -1.034317017673926
PM25 2023
-21.435097073529267 4.484382050688563 -1.0182999032067261 -1.2947849174331718
PM10 2019
-10.92173244654924 1.5461221262830225 -1.4065824093796813 -1.5081430603553918
PM10 2020
-8.387341515038566 15.030613234629275 -1.232904425838534 -1.3456087603694686
PM10 2021
-8.405029648111697 3.4648176225266245 -1.5175773014380969 -1.60774199310788
PM10 2022
-6.763784865196159 17.084469811304945 -1.3916240293168978 -1.3465895147967464
PM10 2023
-8.693871264325832 17.075256618093913 -1.260581049808478 -1.018895547043378
NO2 2019
-25.835106080532523 30.030419761260774 -1.6421693756613633 -1.2085676285394333
NO2 2020
-18.97554796618971 123.450059168

#
We also find 25 instances across 15 sites where sites become non-compliant with annual NAAQS NO2 standards after data cleaning

In [27]:
smmp = summary_mean_map_plot[summary_mean_map_plot['site_id'].isin(sites_in_consideration['site_id'][sites_in_consideration['pollutant'] == 'NO2'])]
print('NO2')
print(smmp['site_id'].nunique())
print(smmp[smmp['compliance_change_compl_to_non']]['site_id'].nunique())
print()
smmp = summary_mean_map_plot[summary_mean_map_plot['site_id'].isin(sites_in_consideration['site_id'][sites_in_consideration['pollutant'] == 'PM10'])]
print('PM10')
print(smmp['site_id'].nunique())
print(smmp[smmp['compliance_change_compl_to_non']]['site_id'].nunique())
print()
smmp = summary_mean_map_plot[summary_mean_map_plot['site_id'].isin(sites_in_consideration['site_id'][sites_in_consideration['pollutant'] == 'PM25'])]
print('PM25')
print(smmp['site_id'].nunique())
print(smmp[smmp['compliance_change_compl_to_non']]['site_id'].nunique())
print()

smmp = summary_mean_map_plot[summary_mean_map_plot['site_id'].isin(sites_in_consideration['site_id'][sites_in_consideration['pollutant'] == 'Ozone'])]
print('Ozone')
print(smmp['site_id'].nunique())
print(smmp[smmp['compliance_change_compl_to_non']]['site_id'].nunique())

NO2
184
77

PM10
182
74

PM25
193
79

Ozone
188
79


#
After data cleaning, we see an average decrease of 20% in NOx-to-NO2 ratio across the study period, mostly from 2.4-3.2 to 1.6-2.4

In [30]:
NOx_ratio = NO2_sites_mean['NO2']/NO2_sites_mean['NO']
New_NOx_ratio = NO2_sites_mean['NO2_clean']/NO2_sites_mean['NO_clean']
#Percentage Change
((New_NOx_ratio - NOx_ratio)/NOx_ratio * 100).describe()

count    891.000000
mean       5.145754
std       16.427996
min      -77.147373
25%       -0.204976
50%        3.833521
75%        8.930065
max      161.329434
dtype: float64

#
Mumbai, Delhi, Kolkata

In [3]:
df = summary_mean
df['lat'] = df['site_id'].map(site_master.set_index('site_id')['latitude'])
df['lon'] = df['site_id'].map(site_master.set_index('site_id')['longitude'])
mumbai = df[(df['lat'] > 19.03) & (df['lat'] < 19.112) & (df['lon'] > 72.82) & (df['lon'] < 72.9)]
kolkata = df[(df['lat'] > 22.48) & (df['lat'] < 22.662) & (df['lon'] > 88.25) & (df['lon'] < 88.45)]
delhi = df[(df['lat'] > 28.34) & (df['lat'] < 28.88) & (df['lon'] > 76.08) & (df['lon'] < 77.4)]
mumbai['city'] = 'Mumbai'
kolkata['city'] = 'Kolkata'
delhi['city'] = 'Delhi'


C:\Users\hitan\AppData\Local\Temp\ipykernel_23592\804434225.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  mumbai['city'] = 'Mumbai'
C:\Users\hitan\AppData\Local\Temp\ipykernel_23592\804434225.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  kolkata['city'] = 'Kolkata'
C:\Users\hitan\AppData\Local\Temp\ipykernel_23592\804434225.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in 

In [4]:
pd.concat([mumbai, kolkata, delhi], ignore_index=True).to_csv('CPCB_Issues/AirPy_v2/new_data/summary/summary_mean_major_cities.csv', index=False)

{'NO2': [212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254, 255, 256, 257, 258, 259, 260, 261, 262, 263, 264, 265, 266, 267, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277, 278, 279, 280, 281, 282, 283, 284, 285, 286, 287, 288, 289, 290, 291, 292, 293, 294, 295, 296, 297, 298, 299, 300, 301, 302, 303, 304, 305, 306, 307, 308, 309, 310, 311, ...], 'Ozone': [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, ...], 'PM10': [636, 637, 638, 639, 640, 641, 642, 643, 644, 645, 646, 647, 648, 649, 

In [39]:
import pandas as pd
print('2019')
map_plot_summary = pd.read_csv(r'CPCB_Issues\AirPy_v2\new_data\summary\2019_summary_mapping.csv')
groups = map_plot_summary.groupby('pollutant')
# groups.groups
for group in groups.groups:
    # print(groups.get_group(group).columns)
    pol = groups.get_group(group)
    pol = pol[(pol['Ozone_DA'] >= 50) & (pol['Ozone_DA_before'] >= 75)]
    print(f"{group} | MIN: {pol['percentage_change'].min():.2f} | MAX: {pol['percentage_change'].max():.2f}")

print('\n2023')
map_plot_summary = pd.read_csv(r'CPCB_Issues\AirPy_v2\new_data\summary\2023_summary_mapping.csv')
groups = map_plot_summary.groupby('pollutant')
# groups.groups
for group in groups.groups:
    # print(groups.get_group(group).columns)
    pol = groups.get_group(group)
    pol = pol[(pol['Ozone_DA'] >= 50) & (pol['Ozone_DA_before'] >= 75)]
    print(f"{group} | MIN: {pol['percentage_change'].min():.2f} | MAX: {pol['percentage_change'].max():.2f}")

2019
NO2 | MIN: -13.95 | MAX: 96.17
Ozone | MIN: -7.51 | MAX: 5.40
PM10 | MIN: -4.58 | MAX: 0.57
PM25 | MIN: -15.13 | MAX: 0.25

2023
NO2 | MIN: -43.58 | MAX: 113.12
Ozone | MIN: -19.11 | MAX: 35.26
PM10 | MIN: -8.70 | MAX: 13.54
PM25 | MIN: -21.83 | MAX: 4.47
